File: backend/controllers/khachHangController.js
const KhachHangService = require("../services/khachHangService")
const service = new KhachHangService()

// Controller theo sơ đồ lớp
class KhachHangController {
  async getAllKhachHang(req, res) {
    try {
      const khachHangs = await service.getAllKhachHang()
      res.json(khachHangs)
    } catch (error) {
      res.status(500).json({ message: error.message })
    }
  }

  async getKhachHangById(req, res) {
    try {
      const { id } = req.params
      const khachHang = await service.getKhachHangById(Number.parseInt(id))
      if (!khachHang) {
        return res.status(404).json({ message: "Không tìm thấy khách hàng" })
      }
      res.json(khachHang)
    } catch (error) {
      res.status(500).json({ message: error.message })
    }
  }

  async createKhachHang(req, res) {
    try {
      const newKhachHang = await service.createKhachHang(req.body)
      res.status(201).json(newKhachHang)
    } catch (error) {
      res.status(400).json({ message: error.message })
    }
  }

  async updateKhachHang(req, res) {
    try {
      const { id } = req.params
      const updatedKhachHang = await service.updateKhachHang(Number.parseInt(id), req.body)
      if (!updatedKhachHang) {
        return res.status(404).json({ message: "Không tìm thấy khách hàng" })
      }
      res.json(updatedKhachHang)
    } catch (error) {
      res.status(400).json({ message: error.message })
    }
  }

  async deleteKhachHang(req, res) {
    try {
      const { id } = req.params
      const result = await service.deleteKhachHang(Number.parseInt(id))
      if (!result) {
        return res.status(404).json({ message: "Không tìm thấy khách hàng" })
      }
      res.json({ message: "Xóa khách hàng thành công" })
    } catch (error) {
      res.status(500).json({ message: error.message })
    }
  }
}

module.exports = new KhachHangController()

File: customer-service/src/main/java/com/example/customerservice/controller/KhachHangController.java
package com.example.customerservice.controller;

import com.example.customerservice.model.KhachHang;
import com.example.customerservice.service.KhachHangService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/khach-hang")
public class KhachHangController {

    private final KhachHangService khachHangService;

    public KhachHangController(KhachHangService khachHangService) {
        this.khachHangService = khachHangService;
    }

    @GetMapping
    public ResponseEntity<List<KhachHang>> getAllKhachHangs() {
        List<KhachHang> khachHangs = khachHangService.getAllKhachHangs();
        return ResponseEntity.ok(khachHangs);
    }

    @GetMapping("/{id}")
    public ResponseEntity<KhachHang> getKhachHangById(@PathVariable Long id) {
        KhachHang khachHang = khachHangService.getKhachHangById(id);
        return ResponseEntity.ok(khachHang);
    }

    @GetMapping("/theo-chi-tieu")
    public ResponseEntity<List<KhachHang>> getKhachHangsByTongChiTieu() {
        List<KhachHang> khachHangs = khachHangService.getKhachHangsByTongChiTieu();
        return ResponseEntity.ok(khachHangs);
    }

    @PostMapping
    public ResponseEntity<KhachHang> createKhachHang(@RequestBody KhachHang khachHang) {
        KhachHang createdKhachHang = khachHangService.createKhachHang(khachHang);
        return new ResponseEntity<>(createdKhachHang, HttpStatus.CREATED);
    }

    @PutMapping("/{id}")
    public ResponseEntity<KhachHang> updateKhachHang(@PathVariable Long id, @RequestBody KhachHang khachHang) {
        KhachHang updatedKhachHang = khachHangService.updateKhachHang(id, khachHang);
        return ResponseEntity.ok(updatedKhachHang);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteKhachHang(@PathVariable Long id) {
        khachHangService.deleteKhachHang(id);
        return ResponseEntity.noContent().build();
    }
}
